File: problem/normal/寻找两个有序数组的中位数/code/solution_1.py
def find_median_sorted_arrays(nums1, nums2):
    m, n = len(nums1), len(nums2)
    total = m + n
    if total % 2 == 1:
        k = total // 2
        i, j = 0, 0
        while k > 0:
            if i < m and (j >= n or nums1[i] <= nums2[j]):
                i += 1
            else:
                j += 1
            k -= 1
        return float(min(nums1[i], nums2[j]))
    else:
        k1, k2 = total // 2 - 1, total // 2
        i, j = 0, 0
        a, b = 0, 0
        while k1 > 0:
            if i < m and (j >= n or nums1[i] <= nums2[j]):
                a = nums1[i]
                i += 1
            else:
                a = nums2[j]
                j += 1
            k1 -= 1
        if i < m and (j >= n or nums1[i] <= nums2[j]):
            b = nums1[i]
        else:
            b = nums2[j]
        return (a + b) / 2.0

if __name__ == "__main__":
    import sys
    input = sys.stdin.read().split()
    ptr = 0
    m, n = int(input[ptr]), int(input[ptr+1])
    ptr += 2
    nums1 = list(map(int, input[ptr:ptr+m]))
    ptr += m
    nums2 = list(map(int, input[ptr:ptr+n]))
    print("{0:.1f}".format(find_median_sorted_arrays(nums1, nums2)))

File: problem/normal/寻找两个有序数组的中位数/code/solution_1.java
import java.util.*;

public class Main {
    public static double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int m = nums1.length, n = nums2.length;
        int total = m + n;
        if (total % 2 == 1) {
            int k = total / 2;
            int i = 0, j = 0;
            while (k-- > 0) {
                if (i < m && (j >= n || nums1[i] <= nums2[j])) {
                    i++;
                } else {
                    j++;
                }
            }
            return Math.min(nums1[i], nums2[j]);
        } else {
            int k1 = total / 2 - 1, k2 = total / 2;
            int i = 0, j = 0;
            int a = 0, b = 0;
            while (k1-- > 0) {
                if (i < m && (j >= n || nums1[i] <= nums2[j])) {
                    a = nums1[i];
                    i++;
                } else {
                    a = nums2[j];
                    j++;
                }
            }
            if (i < m && (j >= n || nums1[i] <= nums2[j])) {
                b = nums1[i];
            } else {
                b = nums2[j];
            }
            return (a + b) / 2.0;
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int m = scanner.nextInt();
        int n = scanner.nextInt();
        int[] nums1 = new int[m];
        int[] nums2 = new int[n];
        for (int i = 0; i < m; i++) nums1[i] = scanner.nextInt();
        for (int i = 0; i < n; i++) nums2[i] = scanner.nextInt();
        System.out.printf("%.1f\n", findMedianSortedArrays(nums1, nums2));
    }
}